src/ovs/aca_ovs_l3_programmer.cpp

```diff
@@ -42,49 +42,346 @@ ACA_OVS_L3_Programmer &ACA_OVS_L3_Programmer::get_instance()
+            throw std::invalid_argument("'/' not found in cidr");
+          }
+
+          found_tunnel_id = current_SubnetConfiguration.tunnel_id();
+          if (!aca_validate_tunnel_id(found_tunnel_id)) {
+            throw std::invalid_argument("found_tunnel_id is invalid");
+          }
+
+          // subnet info's gateway ip and mac needs to be there and valid
+          found_gateway_ip = current_SubnetConfiguration.gateway().ip_address();
+
+          // inet_pton returns 1 for success 0 for failure
+          if (inet_pton(AF_INET, found_gateway_ip.c_str(), &(sa.sin_addr)) != 1) {
+            throw std::invalid_argument("found gateway ip address is not in the expect format");
+          }
+
+          found_gateway_mac = current_SubnetConfiguration.gateway().mac_address();
+
+          if (!aca_validate_mac_address(found_gateway_mac.c_str())) {
+            throw std::invalid_argument("found_gateway_mac is invalid");
+          }
+
+          subnet_routing_table_entry new_subnet_routing_table_entry;
+
+          if (is_subnet_routing_table_exist) {
+            new_subnet_routing_table_entry =
+                    new_subnet_routing_tables[current_router_subnet_id];
+          }
+
+          // update the subnet routing table entry
+          new_subnet_routing_table_entry.vpc_id = found_vpc_id;
+          new_subnet_routing_table_entry.network_type =
+                  current_SubnetConfiguration.network_type();
+          new_subnet_routing_table_entry.cidr = found_cidr;
+          new_subnet_routing_table_entry.tunnel_id = found_tunnel_id;
+          new_subnet_routing_table_entry.gateway_ip = found_gateway_ip;
+          new_subnet_routing_table_entry.gateway_mac = found_gateway_mac;
+          // don't need to handle the gateway_ip and gateway_mac change, because that will
+          // require the subnet to remove the gateway port and add in a new one
+
+          source_vlan_id = ACA_Vlan_Manager::get_instance().get_or_create_vlan_id(found_vpc_id);
+
+          current_gateway_mac = found_gateway_mac;
+          current_gateway_mac.erase(
+                  remove(current_gateway_mac.begin(), current_gateway_mac.end(), ':'),
+                  current_gateway_mac.end());
+
+          addr = inet_network(found_gateway_ip.c_str());
+          snprintf(hex_ip_buffer, HEX_IP_BUFFER_SIZE, "0x%08x", addr);
+
+          // Program ARP responder:
+          cmd_string = "add-flow br-tun \"table=51,priority=50,arp,dl_vlan=" +
+                       to_string(source_vlan_id) + ",nw_dst=" + found_gateway_ip +
+                       " actions=move:NXM_OF_ETH_SRC[]->NXM_OF_ETH_DST[],mod_dl_src:" + found_gateway_mac +
+                       ",load:0x2->NXM_OF_ARP_OP[],move:NXM_NX_ARP_SHA[]->NXM_NX_ARP_THA[],move:NXM_OF_ARP_SPA[]->NXM_OF_ARP_TPA[],load:0x" +
+                       current_gateway_mac +
+                       "->NXM_NX_ARP_SHA[],load:" + string(hex_ip_buffer) +
+                       "->NXM_OF_ARP_SPA[],in_port\"";
```

Comment:
This rule for ARP responder is not simple at all. Please consider to standardize the rule generation by adding a common util method, or one step further, make it part of a new rule generator class in ACA (with input like rule type, rule parameter etc.)

Reply:
yes, that will happen when we covert execute_ovsdb_command and execute_openflow_command shell calls to openflow protocol issue #122